remme/ws/events.py

```diff
@@ -0,0 +1,237 @@
+        try:
+            resp = self._socket.recv_multipart(flags=zmq.NOBLOCK)[-1]
+        except zmq.Again as e:
+            LOGGER.info("No message received yet")
+            return
+
+        # Parse the message wrapper
+        msg = Message()
+        msg.ParseFromString(resp)
+
+        LOGGER.info(f"message type {msg.message_type}")
+
+        # Validate the response type
+        if msg.message_type != Message.CLIENT_EVENTS:
+            LOGGER.info("Unexpected message type")
+            return
+
+        # Parse the response
+        event_list = EventList()
+        event_list.ParseFromString(msg.content)
+
+        web_socks_to_notify = {}
+        LOGGER.info(f"Received events: {event_list.events}")
+        block_num = None
+        block_id = None
+        is_event_catch_up = False
+        for event in event_list.events:
+            # assumed to be the first one in the list
+            if event.event_type == SAWTOOTH_BLOCK_COMMIT_EVENT_TYPE:
+                block_num = int(get_value_from_key(event.attributes, "block_num"))
+                block_id = get_value_from_key(event.attributes, "block_id")
+                is_event_catch_up = block_num <= self.last_block_num
+                self.last_block_num = max(self.last_block_num, block_num)
+            else:
+                event_response = {}
+                event_response['type'] = event.event_type
+                event_response['transaction_id'], event_response['data'] = process_event(event.event_type, event.attributes)
+
+                event = self._events[event_response['type']]
+                for web_sock in event:
+                    if web_sock.closed:
+                        await self._handle_unsubscribe(web_sock)
+                        continue
+
+                    if is_event_catch_up and not event[web_sock]['is_catch_up']:
+                        continue
+
+                    if web_sock not in web_socks_to_notify:
+                        web_socks_to_notify[web_sock] = []
+
+                    web_socks_to_notify[web_sock] += [event_response]
+                    if self.last_block_num == block_num:
+                        self._events[event_response['type']][web_sock]['is_catch_up'] = False
+
+        for web_sock, events in web_socks_to_notify.items():
+            await self._ws_send_message(web_sock, {Entity.EVENTS.value: events,
+                                                   "block_num": block_num,
+                                                   "block_id": block_id,
+                                                   "last_known_block_num": self.last_block_num})
+
+    async def listen_events(self, delta=5):
+        while True:
+            LOGGER.debug('Start events fetching...')
+            await asyncio.gather(*[self.check_event()])
+            await asyncio.sleep(delta)
+
+    def _make_subscriptions(self):
+        return [EventSubscription(event_type=event_name) for event_name in self._events] + \
```

Comment:
Why not add to the self._events?

Reply:
Because self._events is used for string comparison